driver/src/orderbook/streamed/state.rs

```diff
@@ -158,11 +177,150 @@ impl State {
         Ok(())
     }
 
-    fn order_deletion(&mut self, event: &OrderDeletion) -> Result<(), Error> {
+    fn order_deletion(
+        &mut self,
+        event: &OrderDeletion,
+        block_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        if let Some(order) = self.orders.get(&(event.owner, event.id)) {
+            if order.is_valid_in_batch(block_batch_id - 1) {
+                return Err(Error::DeletingValidOrder);
+            } else {
+                self.orders.remove(&(event.owner, event.id));
+            }
+        }
         // Orders are allowed to be deleted multiple times.
-        self.orders.remove(&(event.owner, event.id));
         Ok(())
     }
+
+    fn trade(&mut self, event: &Trade) -> Result<(), Error> {
+        if [event.sell_token, event.buy_token]
+            .iter()
+            .any(|token| self.tokens.get_address_by_id(*token).is_none())
+        {
+            return Err(Error::UnknownToken);
+        }
+        if self.orders.get(&(event.owner, event.order_id)).is_none() {
+            return Err(Error::UnknownOrder);
+        }
+        match &mut self.pending_solution {
+            PendingSolution::AccumulatingTrades(trades) => trades.push(event.clone()),
+            // If there has been a solution before and we get a new trade then the previous solution
+            // must have been reverted.
+            PendingSolution::Submitted { .. } => {
+                self.pending_solution = PendingSolution::AccumulatingTrades(vec![event.clone()])
+            }
+        };
+        Ok(())
+    }
+
+    fn solution_submission(
+        &mut self,
+        event: &SolutionSubmission,
+        block_batch_id: BatchId,
+    ) -> Result<(), Error> {
+        if self.tokens.get_address_by_id(0).is_none() {
+            return Err(Error::SolutionWithoutFeeToken);
+        }
+        let trades = match &mut self.pending_solution {
+            PendingSolution::AccumulatingTrades(trades) => std::mem::replace(trades, Vec::new()),
+            // This would be weird because it means that the previous solution had no trades. Not
+            // sure if we should error.
```

Comment:
Yes, as that would mean it is a solution with 0 utility which is not allowed and something went wrong.